### DISCUSSION TITLE
GDScript pipe operator |>

### DISCUSSION THREAD

**Discussion by wyvernbw** [godotengine/godot-proposals#6481] [Scripting]
I think a pipe operator in gdscript could greatly improve readability of code. For those who don't know, the pipe operator allows you to pass the return value of a function as an argument to another function, something like this:
```elixir
foo(bar(baz(new_function(other_function()))))
# vs
other_function() |> new_function() |> baz() |> bar() |> foo()
```
Imo this is way cleaner and I think gdscript being focused on simplicity as a scripting language can benefit from something like this.

Piping multiple values into a function is forbidden in languages like elixir:
```sh
warning: parentheses are required when piping into a function call.
For example:

  foo 1 |> bar 2 |> baz 3

is ambiguous and should be written as

  foo(1) |> bar(2) |> baz(3)
```
Curios if anybody else might find this useful

**Comment by AThousandShips** [comment]
Given how limited this is, only a single argument, and being unable to call functions that have more than one argument, this seems a very roundabout way to do the same thing, and I don't think it provides any more readability

**Reply by vonagam** [reply]
Just wanted to point out that `unable to call functions that have more than one argument` is not correct.

As for an example of better readability - with pipe operator it is much easier to introduce and use your own helpers for built-in types:

```gdscript
some_array
|> MyArray.map_with_index(func(element, index): return index + ": " + element)
|> MyArray.join(", ")
```
Array in Godot does not have methods like `map_with_index` or `join` - with this you can write them and then organically chain and combine them.

**Comment by wyvernbw** [comment]
It doesn't have to be exactly like elixir. The hack programming language doesn't forbid multiple arguments. We can have a symbol that denotes where the argument gets inserted (kind of like format strings). It can also be used in constructors and array indexing or anything like that
```gdscript
get_something() |> is_equal_approx(%, 0.0) 

var vec = get_something() |> Vector2(10.0, %)

var element = get_index() |> array[%] 
```

**Comment by SKFrozenCloud** [comment]
I like the idea, since in functional programming it is used a lot. Sure if you don't write in functional programming style you may find it not useful but people who write stuff in that style, it is very useful.

Furthermore, you can store the new composed lambda function in a variable and call that function with any data you want.

**Comment by ShalokShalom** [comment]
The library `golden gadget` adds a couple of features from functional languages and amongst those, also pipe.

Its implemented as a method, so you dont use the pipe operator, but it has the same functionality.

https://monnef.gitlab.io/golden-gadget/features#pipe
